[DCP] OSS Zero Overhead Checkpointing Implementation

Saiteja64 opened the pull request:
Summary: This diff updates DCP driver code/APIs to support Zero Overhead Checkpointing

Test Plan: Test with TorchTitan on this PR: https://github.com/pytorch/torchtitan/pull/1287

Differential Revision: D72391401




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

vadimkantorov commented:
Does the async DCP also work/practical for sending checkpoints directly to s3? (I guess using https://github.com/awslabs/s3-connector-for-pytorch ? or maybe is there some native support?)

fegin commented on torch/distributed/_pin_memory_utils.py:
```diff
@@ -0,0 +1,66 @@
+logger.setLevel(logging.INFO)
+
+
+# Allows retrying cudaHostRegister if it fails
+CKPT_PIN_ALLOW_RETRY = os.environ.get("CKPT_PIN_ALLOW_RETRY", "1") == "1"
+# Peeks last cudaError before pinning shared memory
+CKPT_PIN_PEEK_CUDA_ERROR = os.environ.get("CKPT_PIN_PEEK_CUDA_ERROR", "0") == "1"
+# Pops last cudaError before pinning shared memory
+CKPT_PIN_POP_CUDA_ERROR = os.environ.get("CKPT_PIN_POP_CUDA_ERROR", "0") == "1"
```
Do we want to export these variables? I guess not. We can just move before where they are used. For example, `CKPT_PIN_ALLOW_RETRY` is not used at all.

vadimkantorov replied:
Also, should these be `TORCH_`-prefixed?

fegin commented on torch/distributed/_pin_memory_utils.py:
```diff
@@ -0,0 +1,66 @@
+# mypy: allow-untyped-defs
```
Should we have a unittest for this file if possible?

fegin commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -19,6 +25,16 @@
+logger = getLogger()
+logger.setLevel(logging.INFO)
+
+# Allows retrying cudaHostRegister if it fails
+CKPT_PIN_ALLOW_RETRY = os.environ.get("CKPT_PIN_ALLOW_RETRY", "1") == "1"
+# Peeks last cudaError before pinning shared memory
+CKPT_PIN_PEEK_CUDA_ERROR = os.environ.get("CKPT_PIN_PEEK_CUDA_ERROR", "0") == "1"
+# Pops last cudaError before pinning shared memory
+CKPT_PIN_POP_CUDA_ERROR = os.environ.get("CKPT_PIN_POP_CUDA_ERROR", "0") == "1"
```
ditto. Should just move to right before they are used and only get the environment variables that are actually used.

fegin commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -83,6 +106,7 @@ def _iterate_state_dict(
     ranks_only: tuple[int, ...] = (),
     type_check: bool = True,
     non_blocking: bool = True,
+
```
Remove this empty line.

fegin commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -114,10 +138,12 @@ def _iterate_state_dict(
     ):
         ret = iter_object
+    elif isinstance(iter_object, io.BytesIO):
+        ret = copy.deepcopy(iter_object)
```
We should not do deepcopy as this violates the semantic of `_iterate_state_dict`.

fegin commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -181,6 +189,36 @@ def save(
+def close(): 
+    """
+        MUST CALL THIS FUNCTION AFTER TRAINING IS COMPLETE IF USING ASYNC_SAVE
+    """
```
Probably not a good idea to ask users to selectively call `close()`. I would suggest that users should always call `close()`. Also if this is a public API, the docstring should follow the template. You can check other docstring.

fegin replied:
Should we also wait for the last `async_save` inside this API as well?

Saiteja64 replied:
I think we should just leave that to the users. In general, I want to limit global training state within DCP as much as possible.

fegin commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -181,6 +189,36 @@ def save(
+    staging_completion: Future[None]
+    upload_completion: Future[None]
+
+def close(): 
```
Add return value, `->None:`.

fegin commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -23,18 +27,22 @@
 
 __all__ = ["save_state_dict", "save", "async_save", "AsyncCheckpointerType"]
 
+CACHED_STATE_DICT: Optional[STATE_DICT_TYPE] = None
```
We are not going to expose this variable to users, so we need to make it a private variable.

fegin commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -181,6 +189,36 @@ def save(
+    return staged_state_dict
+
+@dataclass
+class AsyncSaveResponse:
```
Need a good docstring since this is a public dataclass.

fegin commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -181,6 +189,36 @@ def save(
             planner=planner,
         )
 
+def default_stage(state_dict: STATE_DICT_TYPE, cpu_state_dict: STATE_DICT_TYPE) -> STATE_DICT_TYPE:
```
Make this a private function.

fegin reviewed (dismissed):
A general question, what if users mix `async_save()` with `save()`, does this PR handle this case?

fegin commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -249,49 +305,70 @@ def async_save(
-
-    maybe_synchronize_staging()
-
-    return f
```
It will be nice that both paths return `AsyncSaveResponse` -- one path has only `upload_future` and another path has both. But I understand that this will break BC. Not sure if we can do some tricks, like making `AsyncSaveResponse` inherit from `Future`. Just a thought, not necessarily work.

Saiteja64 replied:
yeah it's unfortunate but don't think there is a clean way to do this :/

fegin commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -19,6 +23,8 @@
 
 
+logger = getLogger()
+logger.setLevel(logging.INFO)
```
We should remove this line. Users should be able to control the logging level, not the module.

fegin reviewed (commented):
Overall, looks good. We should remove set logging level. Also, we should have at least one unittest for this feature.

teja-rao reviewed (commented):
I added a few comments but i think i have a fundamental question: Should we support zero overhead copy in async stager implementation instead in the legacy staging path?

teja-rao commented on torch/distributed/checkpoint/_async_executor.py:
```diff
@@ -15,7 +15,7 @@ class _AsyncCheckpointExecutor(abc.ABC):
     def execute_save(
         self,
-        staged_state_dict: STATE_DICT_TYPE,
+        staged_state_dict_future: Future[STATE_DICT_TYPE],
```
should this be STATE_DICT_TYPE | Future[STATE_DICT_TYPE]?

nit: i see we are converting the state_dict_type to future to pass in which is okay but i think it is more readable to just keep passing in the Union of state_dict_type and future.

Saiteja64 replied:
That makes sense. I was originally thinking in the long term, stage would only return a future (would work for sync as well) but I don't see that happening because it would introduce breaking changes

teja-rao commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -249,49 +296,65 @@ def async_save(
             "A CPU backend must be enabled for async save; try initializing process group with 'cpu:gloo,cuda:nccl'"
         )
 
+    if async_stager is None:
+        if (storage_writer is None or not isinstance(storage_writer, AsyncStager)):
+            async_stager = DefaultStager(StagingOptions(not block_on_staging, not block_on_staging, not block_on_staging, not block_on_staging))
+        elif isinstance(storage_writer, AsyncStager):
+            # bwc with old storage_writers
+            async_stager = storage_writer
+
```
nit:
```suggestion
    if async_stager is None:
        if (storage_writer is not None and isinstance(storage_writer, AsyncStager)):
                    # bwc with old storage_writers
                   async_stager = storage_writer
        else:
                    async_stager = DefaultStager(StagingOptions(not block_on_staging, not block_on_staging, not block_on_staging, not block_on_staging))

        

```

teja-rao commented on torch/distributed/checkpoint/_async_thread_executor.py:
```diff
@@ -11,24 +11,41 @@
         f: Future = self._executor.submit(
-            save,
-            staged_state_dict,
+            save_wrapper,
```
is save method still used for sync save? why not change it to support union?

teja-rao reviewed (commented):
overall pr looks good to me after updates. I will stamp the PR once the testing is complete and CI shows green! Thank you for making the changes.

Saiteja64 commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -191,12 +210,15 @@ def async_save(
     process_group: Optional[dist.ProcessGroup] = None,
     async_checkpointer_type: AsyncCheckpointerType = AsyncCheckpointerType.THREAD,
-) -> Future:
+    async_stager: Optional[AsyncStager] = None,
+    block_on_staging: bool = True,
```
For async_stager argument, this is a directional change. We are separating out the stager from storage_writer. Conceptually, you should be able to use a stager with any storage_writer. Combining the two just forces us and users to change StorageWriter every time we want to make a change to stager (new properties, etc)

block_on_staging: Think we can remove this one. Works for all cases and simplifies API:
1. If user configures AsyncStager to return a future instead of a state_dict, driver code can just return that future. 
2. If AsyncStager returns state_dict we can just keep existing logic of checking synchronize_after_execute and call synchronize_staging if needed

teja-rao commented on torch/distributed/checkpoint/_checkpointer.py:
```diff
@@ -83,12 +83,14 @@ def async_save(
             process_group=self.process_group,
             planner=self.save_planner,
         )
+        assert isinstance(response, Future)
```
is this assert needed? mypy typechecks should catch if you arent returning a future?

Saiteja64 replied:
Without this, we introduce a linter error because async_save either returns a Tuple of staging_future/upload_future or an upload future now.

teja-rao commented on torch/distributed/checkpoint/staging.py:
```diff
@@ -1,11 +1,31 @@
+- Caching of frequently used storages for efficient memory management
+- Automatic resource cleanup and memory management
+Classes:
+    CheckpointStager: Abstract base class defining the staging interface
+    StagingOptions: Configuration dataclass for staging behavior
```
clean up/update? i think these are from dcp evolution work..

teja-rao commented on torch/distributed/checkpoint/staging.py:
```diff
@@ -56,12 +77,183 @@ def stage(self, state_dict: STATE_DICT_TYPE) -> STATE_DICT_TYPE:
+        staged_dict = future.result()
+        stager.close()
+
+        # Context manager pattern (recommended)
+        with DefaultStager(config) as stager:
+        result = stager.stage(state_dict)
```
i think we do not want users to create a stager each time. stager caches the storages, may be this needs an update.

teja-rao commented on torch/distributed/checkpoint/staging.py:
```diff
@@ -56,12 +77,183 @@ def stage(self, state_dict: STATE_DICT_TYPE) -> STATE_DICT_TYPE:
+        if self._staging_executor:
+            self._staging_executor.shutdown(wait=True)
+
+    def synchronize_staging(self) -> None:
```
throw an exception and suggest synchronizing using the future or call staging_future.result() here?

teja-rao reviewed (changes requested):
sending back for updating docs and for consideration on nits.

teja-rao reviewed (approved):
approving to unblock, please fix the mypy error before landing.

Saiteja64 committed:
[DCP] OSS Zero Overhead Checkpointing Implementation (#156207)

Summary:
X-link: https://github.com/pytorch/tnt/pull/1010


This diff updates DCP driver code/APIs to support Zero Overhead Checkpointing

Test Plan:
Test with TorchTitan on this PR: https://github.com/pytorch/torchtitan/pull/1287

Add new UT

Reviewed By: diego-urgell

Differential Revision: D72391401

Saiteja64 commented:
@pytorch merge

Saiteja64 commented:
@pytorchbot merge

Saiteja64 commented:
@pytorchbot merge
